specification/under_development/Compilation_And_Targeting.md

```diff
@@ -0,0 +1,67 @@
+
+What we gain by adding a linking stage with some customizations, compared to just use the standard llvm linker:
+
+- It is still possible to have a simulator (or hardware for that matter) that implements several gate sets but use a specific gate set, i.e. only a subset of the implemented gates, during execution.
+- It avoids forcing different front ends to effectively align on naming convention at the IR level.
+- It permits to check signatures and fail at compile time if they don't match.
+
+### Targeting
+
+The QIR specification defines which instructions, including runtime functions and QIS functions, are sufficient to support arbitrary computations that make use of both quantum and classical resources, and gives recommendations for how to leverage and combine these building blocks in a way that ensures maximal compatibility with different backends. Most or possibly all backends today do not support arbitrary computations. A common requirement for a program to be executable on a quantum backend is, for example, that it is possible to ensure during compilation that the quantum computation terminates. Such requirements poses fundamental limitations for which programs can or cannot be executed on a specific backend. Compiling a program in a way that maximizes compatibility with different backends means that the program representation facilitates that further compilation steps can largely eliminate constructs that are not supported by the backend. This targeting process to meet backend requirements compiles QIR into a QIR profile. It may result in a compilation failure if certain constructs are not supported by the backend and cannot reasonably be eliminated/replaced as part of targeting. For example, suppose the targeted backend requires measurements to be performed only as a final instruction on a QPU, i.e. all quantum resources are released after measurement. If the program contains branching based on measurements, i.e. subsequent quantum computations depend on a prior measurement result, then this construct (meaning the branching) can be eliminate by replacing it with a coherent version according to the [Deferred Measurement Principle](https://en.wikipedia.org/wiki/Deferred_Measurement_Principle). However, this may be impractical since it may require a (potentially significantly) larger number of qubits to perform the computation, such that the targeting stage may choose to fail compilation instead.
+
+The QIR specification outlines how to ensure that certain data structures, control flow structures, and classical computations can be eliminated during targeting. It is up to each front end to determine the implications for when source code can be successfully compiled and executed and whether or not to give early feedback during development. 
```

Comment:
This seems an important step for front end language designers to want to harness QIR in a portable manner. But are front end language designers discouraged from writing compilers that can compile straight to base profile (or future extended profiles) themselves?

Reply:
I think that is a question that may also be worthwhile touching on in the meeting tomorrow.